More documentation (#36)
peopledoc/cabbage: docs/howto.rst

How-to...
=========

Ensure jobs don't run concurrently and run in order
---------------------------------------------------

Let's imagine we have a task like this::

    @task_manager.task(queue="queue")

>     def write_alphabet(floor):

Author: Oh gosh it should NEVER have been called floor anymore ><

>     def write_alphabet(floor):

Author: ```suggestion
    def write_alphabet(letter):
```

        time.sleep(random.random() * 5)
        with open("/tmp/alphabet.txt", "a") as f:

>             f.write(floor)

Author: ```suggestion
            f.write(letter)
```

We write the letter we recieve to a file after wating for a
random time (this is a simplified version of a real
world problem where tasks take a unforseeable amount of time
and share resources like a database).

We call it::

>     write_alphabet.defer("a")

Author: ```suggestion
    write_alphabet.defer(letter="a")
```

>     write_alphabet.defer("b")

Author: ```suggestion
    write_alphabet.defer(letter="b")
```

>     write_alphabet.defer("c")

Author: ```suggestion
    write_alphabet.defer(letter="c")
```

>     write_alphabet.defer("d")

Author: ```suggestion
    write_alphabet.defer(letter="d")
```

> We could expect the following to be written in `pyramid.txt`::

Reviewer: ```suggestion
We could expect the following to be written in `alphabet.txt`::
```

Author: Oh yeah right. I started with a different example, and simplified it :D

    a
    b
    c
    d

And what we find is pretty much like this::

    d
    a
    c
    b

Jobs were taken from the queue in order, but if we have several
workers, they have been launched in parallel and given their duration
is random, the final result pretty much is too.

We can solve this problem by using locks::

>     job_descritption = write_alphabet.configure(lock="/tmp/alphabet.txt")

Reviewer: using a path as lock parameter could be misleading no ? (even if it's the name of the resource, the task might have multiple resources used)

Author: Not sure what you mean... What do you suggest ?

Note: the whole task makes no sense in a distributed context, the file will be split between all the machines that run the workers.

Reviewer: ok, keep this :)

    job_descritption.defer("a")
    job_descritption.defer("b")
    job_descritption.defer("c")
    job_descritption.defer("d")

In this case, our jobs might still be executed by any of the workers,
but cabbage will not select a job for completion as long as there is

> a job currently processing with the same log. Note that cabbage will

Author: ```suggestion
a job currently processing with the same lock. Note that cabbage will
```

use postgres to search the jobs table for suitable jobs, meaning that

> even if the database contains a high proportion of locked tasks, it wont

Author: ```suggestion
even if the database contains a high proportion of locked tasks, it will barely
```

> affect cabbages's capacity to execute the free tasks. Also, identical

Author: ```suggestion
affect cabbages's capacity to quickly find the free tasks. Also, identical
```

jobs will always be started in creation order, so we can be assured our
tasks will run sequentially and in order.

A good string identifier for the lock is a string identifier of

> the shared resource, UUIDs are well suited for this.

Author: ```suggestion
the shared resource, UUIDs are well suited for this.
If multiple resources are implicated, a combination
of their identifiers could be used (there's no hard
limit on the length of a lock string, but stay reasonable)
```

A task can only take a single lock so there's no dead-lock scenario possible
where 2 running tasks are waiting one another.

There is no mechanism in place to expire locks yet, but if a task fails

> without the whole Python process to crash, it will free its lock.

Author: ```suggestion
without the whole Python process crashing, it will free its lock.
```

Launch a job in the future
---------------------------

If a job is configured with a date in the future, it will run at the
first opportunity after that date. Let's launch the task that will
trigger the infamous 2018 bug::

    dt = datetime.datetime(2038, 1, 19, 3, 14, 7).replace(
        tzinfo=datetime.timezone.utc
    )
    create_bug.configure(schedule_at=dt).defer(crash_everything=True)

Also, you can configure a delay from now::

    clean.configure(schedule_in={"hours": 1, "minutes": 30}).defer()

The details on the parameters you can use are in the pendulum
`documentation <https://pendulum.eustace.io/docs/#addition-and-subtraction>`_
(because we use pendulum under the hood).

Add a task middleware
---------------------

> There's nothing yet do this, but given you control the decorator

Reviewer: I don't understand the first sentence :)
